### PR TITLE
Add 3rd party integration guidelines

### PR DESCRIPTION
**Pull request by jasti**
Suggestions appreciated!

### CODE REVIEW

**Comment by jasti** [comment]
I signed it!

File: ads/README.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # Integrating ad networks into AMP
 
 See also our [ad integration guidelines](../3p/README.md#ads).
+See also our [3rd party ads integration guidelines](/3p.md)
```
Comment (jmadler) [inline ads/README.md]: Can we compress this into one sentence?

File: ads/README.md [hunk 2]
```diff
@@ -159,3 +160,5 @@ Technically the `<amp-ad>` tag loads an iframe to a generic bootstrap URL that k
 
 Access to a publishers 1st party cookies may be achieved through a custom ad bootstrap
 file. See ["Running ads from a custom domain"](../builtins/amp-ad.md) in the ad documentation for details.
+
+If the publisher would like to give custom javascript in the remote.html, the ability to read or write to  the publisher owned cookies, the publisher needs to ensure that the remote.html file is hosted on a sub-domain of the publisher URL. e.g. if the publisher hosts a webpage on https://nytimes.com, then the remote file should be hosted on something similar to https://sub-domain.nytimes.com for the custom JS to have the abiity to read or write to nyt’s cookies.
```
Comment (jmadler) [inline ads/README.md]: s/nyt's cookies/cookies for nytimes.com/

File: ads/3p.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
+### Instructions to integrate, if you are:
```
Comment (jmadler) [inline ads/3p.md]: Replace this header with a quick overview.  e.g.:

If you are an ad technology provider looking to integrate with AMP HTML, please see the following guidelines:

File: ads/3p.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+#### Ad Server
+
+*Examples : DFP, A9*
+
+As an adserver, publishers you support include a javascript library provided by you and place various ‘ad snippets’ that rely on the javascript library to fetch ads and render them on the publisher’s website.
+Since AMP doesn’t allow publishers to execute arbitrary javascript (like the library above), you will need to contribute to the AMP open source code to allow the amp-ad built-in tag to request ads from your ad server.
+
+For example : Amazon A9 server (documentation) can be invoked using following syntax.
```
Comment (jmadler) [inline ads/3p.md]: linkify "documentation"

File: ads/3p.md [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+```
+<amp-ad width=300 height=250
+    type="a9"
+    data-aax_size="300x250"
+    data-aax_pubname="test123"
+    data-aax_src="302">
+</amp-ad>
+```
+
+Also note that each of the attributes that follow ‘type’ are dependent on the parameters that Amazon’s A9 server expects in order to deliver an ad. The file a9.js shows you how the parameters are mapped to making a javascript call which will invoke the server by invoking the URL : https://c.amazon-adsystem.com/aax2/assoc.js  and appending the corresponding parameters being passed in by the AMP ad tag  and return the ad.
```
Comment (jmadler) [inline ads/3p.md]: linkify a9.js

File: ads/3p.md [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+
+In order to ensure minimum latency and quality, please follow the instructions listed here before submitting a pull request to the AMP open source project. For general guidance on how get started with contributing to the AMP project, please see here.
```
Comment (jmadler) [inline ads/3p.md]: linkify, and move to the top

File: ads/3p.md [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+##### Side Platform (SSP) or an Ad Exchange
+
+*Examples : Rubicon, Criteo OR Appnexus, Ad-Exchange*
+
+If you are a sell side platform that wants to directly get called from a publisher’s webpage, you will need to follow the same directions as listed above for integrating with an Ad Server. Adding your own ‘type’ value to the amp-ad tag will allow you to distribute your tag directly to the publisher, so they can insert your tags directly into their AMP pages.
```
Comment (jmadler) [inline ads/3p.md]: remove "as listed above" so we can reorder if we want to :)

File: ads/3p.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+<amp-brightcove
+      data-account="906043040001"
+      data-video-id="1401169490001"
+      data-player-id="180a5658-8be8-4f33-8eba-d562ab41b40c"
+      layout="responsive" width="480" height="270">
+  </amp-brightcove>
+```
+For instructions on how to develop an amp tag like brightcove, please see [amp-brightcove.js](https://github.com/ampproject/amphtml/blob/master/extensions/amp-brightcove/0.1/amp-brightcove.js)
+
+The other popular option is for video players to take advantage of the amp-iframe (https://github.com/ampproject/amphtml/blob/master/extensions/amp-iframe/amp-iframe.md) extension (with post frame or fallback markup) and serve the player within the iframe.
```
Comment (jmadler) [inline ads/3p.md]: let's steer away from amp-iframe for 3rd parties

File: ads/3p.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+You can use a similar approach to enrich the ad call by passing in audience segments that you get from the user cookie into the ad call.
+
+#### Viewability Provider
+
+*Examples : MOAT, Integral Ad Science*
+
+Viewability providers typically integrate with publishers via the ad server’s creative wrappers. If that is the case, ensure that the creative wrapper loads all assets over HTTPS.
+
+For e.g. for MOAT, make sure http://js.moatads.com is switched to  https://z.moatads.com
```
Comment (jmadler) [inline ads/3p.md]: is "js" -> "z" intentional?

File: ads/3p.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+If you instead would like to directly embed viewability JS on the publisher page, use the approach to inject javascript on a remote server. [See how to enhance custom ad configuration](https://github.com/ampproject/amphtml/blob/master/builtins/amp-ad.md#enhance-incoming-ad-configuration)
+
+Alternatively, see appraoch to using the [intersection obeserver pattern](#ad-viewability)
```
Comment (jmadler) [inline ads/3p.md]: misspelled approach

File: ads/3p.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+#### Content-Recommendation Platform
+
+*Examples : Taboola, Outbrain*
+
+As content recommendation platforms, if you have some piece of Javascript embeded on the publisher website today but the approach will not work in AMP pages. If you would like to recommend content via the publisher’s ad server, we suggest that you use the amp-embed extension to request the content details. (served like the creative).
+
+Instead, you can also create your own AMP extension tag which will allow you to embed a tag like <amp-taboola> in the publisher’s page instead of custom javascript.
```
Comment (jmadler) [inline ads/3p.md]: I think we prefer amp-embed / amp-ad over a provider specific tag

**Comment by jmadler** [comment]
Made some suggestions.  Can you include sample PRs where appropriate too?

File: ads/3p.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
+### Instructions to integrate, if you are:
+
+#### Ad Server
+
+*Examples : DFP, A9*
+
+As an adserver, publishers you support include a javascript library provided by you and place various ‘ad snippets’ that rely on the javascript library to fetch ads and render them on the publisher’s website.
```
Comment (cramforce) [inline ads/3p.md]: s/javascript/JavaScript

File: ads/3p.md (hunk 3 above)
Comment (cramforce) [inline ads/3p.md]: Maybe call the file `integration-guide.md`?

File: ads/3p.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+Since AMP doesn’t allow publishers to execute arbitrary javascript (like the library above), you will need to contribute to the AMP open source code to allow the amp-ad built-in tag to request ads from your ad server.
+
+For example : Amazon A9 server (documentation) can be invoked using following syntax. 
+```
```
Comment (cramforce) [inline ads/3p.md]: Make this "```html" for better syntax highlighting.

File: ads/3p.md (hunk 7 above)
Comment (cramforce) [inline ads/3p.md]: "wants to get called directly"

File: ads/3p.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+In order to ensure minimum latency and quality, please follow the instructions listed here before submitting a pull request to the AMP open source project. For general guidance on how get started with contributing to the AMP project, please see here.
+
+##### Side Platform (SSP) or an Ad Exchange
+
+*Examples : Rubicon, Criteo OR Appnexus, Ad-Exchange*
+
+If you are a sell side platform that wants to directly get called from a publisher’s webpage, you will need to follow the same directions as listed above for integrating with an Ad Server. Adding your own ‘type’ value to the amp-ad tag will allow you to distribute your tag directly to the publisher, so they can insert your tags directly into their AMP pages.
+
+More commonly, SSPs work with the publisher to traffick the SSP’s ad tags in their ad server. In this case, ensure that all assets being loaded by your script in the ad server’s creative are being made over HTTPS. There are some restrictions around some intrusive ad formats like expandables, so we recommend that you test out the most commonly delivered creative formats with your publishers.
```
Comment (cramforce) [inline ads/3p.md]: Remove `intrusive`

File: ads/3p.md [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+#### Ad Agency
+*Examples : Essence, Omnicom*
+
+Work with your publisher to ensure that the creatives you develop are AMP compliant. Since all creatives are served into iframes whose size is determined, when the ad is called, ensure that your creative doesn't try to modify the size of the iframe.
+
+Ensure that all assets that are part of the creative are requested using HTTPS.
+Some ad formats are fully not supported at the moment and we recommend testing the creatives in an AMP environment. Some examples are : Rich Media Expandables, Interstitials, Page Level Ads
```
Comment (cramforce) [inline ads/3p.md]: "not fully supported"

File: ads/3p.md [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+
+#### Video Player
+
+*Examples : Brightcove, Ooyala*
+
+A video player that works in regular HTML pages will not work in AMP and therefore a specific tag needs to be created that will allow the AMP runtime to load your player.
```
Comment (cramforce) [inline ads/3p.md]: Please linkify and maybe also link to their PR.

File: ads/3p.md (hunk 8 above)
Comment (cramforce) [inline ads/3p.md]: Its a good point that we could steer them to `amp-embed` which may be more appropriate.

I think you are right, that for 3p platforms (as opposed to, say, ESPN's video player) we should not recommend amp-iframe. I originally suggested this, but lets drop.

File: ads/3p.md (hunk 10 above)
Comment (cramforce) [inline ads/3p.md]: There is no alternative here. The intersection observer approach is the only way to do it.

File: ads/3p.md (hunk 11 above)
Comment (cramforce) [inline ads/3p.md]: yes.

File: ads/3p.md [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+#### Content-Recommendation Platform
+
+*Examples : Taboola, Outbrain*
+
+As content recommendation platforms, if you have some piece of Javascript embeded on the publisher website today but the approach will not work in AMP pages. If you would like to recommend content via the publisher’s ad server, we suggest that you use the amp-embed extension to request the content details. (served like the creative).
+
+Instead, you can also create your own AMP extension tag which will allow you to embed a tag like <amp-taboola> in the publisher’s page instead of custom javascript.
+
+
+If instead you’d like to simply pass in the elements of the content you’d like to recommend to the publisher’s page (so the publisher can maintain the look and feel), use the approach to inject javascript on a remote server. [See how to enhance custom ad configuration](https://github.com/ampproject/amphtml/blob/master/builtins/amp-ad.md#enhance-incoming-ad-configuration)
```
Comment (cramforce) [inline ads/3p.md]: I don't fully get this. Are we suggesting publishers to put some CSS into their remote.html? Not the worst idea :) We should share that with taboola and outbrain.

**Comment by jasti** [comment]
Thanks for the feedback, @jmadler . Will get to them soon.

File: ads/README.md (hunk 1 above)
Comment (jasti) [inline ads/README.md]: Yes.

File: ads/3p.md (hunk 9 above)
Comment (jasti) [inline ads/3p.md]: Yup. Directly from MOAT.

File: ads/3p.md (hunk 17 above)
Comment (jasti) [inline ads/3p.md]: Yeah, that was the idea but <amp-embed> is a much cleaner approach. Removing this.

**Comment by jasti** [comment]
Comments actioned, please review @jmadler @cramforce 
Thanks for the feedback!

File: ads/integration-guide.md [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+
+#### Viewability Provider
+
+*Examples : MOAT, Integral Ad Science*
+
+Viewability providers typically integrate with publishers via the ad server’s creative wrappers. If that is the case, ensure that the creative wrapper loads all assets over HTTPS.
+
+For e.g. for MOAT, make sure http://js.moatads.com is switched to  https://z.moatads.com
+
+Also see approach to using the [intersection obeserver pattern](#ad-viewability)
```
Comment (jmadler) [inline ads/integration-guide.md]: s/obeserver/observer/

Comment (jmadler) [inline ads/integration-guide.md]: Does this link work?

File: ads/integration-guide.md [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+
+#### Content-Recommendation Platform
+
+*Examples : Taboola, Outbrain*
+
+As content recommendation platforms, if you have some piece of Javascript embeded on the publisher website today but the approach will not work in AMP pages. If you would like to recommend content on an AMP page, we suggest that you use the amp-embed extension to request the content details. Please see the [Taboola](https://github.com/ampproject/amphtml/blob/master/ads/taboola.md) example.
```
Comment (jmadler) [inline ads/integration-guide.md]: I would just strike the first sentence of this paragraph

**Comment by jmadler** [comment]
"javascript" should be "JavaScript" throughout

**Comment by jasti** [comment]
please check @jmadler

**Comment by jasti** [comment]
@jmadler @cramforce Please let me know if I should make any more changes or hope to have it merged so 3rd parties can benefit from it. 
Thanks!

**Comment by jmadler** [comment]
cc @Meggin for review

File: ads/README.md [hunk 20]
```diff
@@ -159,3 +159,5 @@ Technically the `<amp-ad>` tag loads an iframe to a generic bootstrap URL that k
 
 Access to a publishers 1st party cookies may be achieved through a custom ad bootstrap
 file. See ["Running ads from a custom domain"](../builtins/amp-ad.md) in the ad documentation for details.
+
+If the publisher would like to give custom JavaScript in the remote.html, the ability to read or write to  the publisher owned cookies, the publisher needs to ensure that the remote.html file is hosted on a sub-domain of the publisher URL. e.g. if the publisher hosts a webpage on https://nytimes.com, then the remote file should be hosted on something similar to https://sub-domain.nytimes.com for the custom JS to have the abiity to read or write cookies for nytimes.com.
```
Comment (cramforce) [inline ads/README.md]: The first sentence has a word missing.

Comment (cramforce) [inline ads/README.md]: `remote.html` should be quoted with ``

**Comment by cramforce** [comment]
Before we can merge, please squash your commits.

**Comment by jasti** [comment]
Closing this PR in favor of #1857 because squashing commits was tricky across file renames. Apologies for the trouble. All changes have been incorporated into #1857. Thanks!

**Pull request closed without merging**